node/src/components/linear_chain_sync.rs

```diff
@@ -367,15 +467,15 @@ where
                     Some(peer) => fetch_block_by_hash(effect_builder, peer, block_hash),
                 },
                 Some(FetchResult::FromStorage(block)) => {
-                    self.state.block_downloaded(&block);
-                    // We should be checking the local storage for linear blocks before we start
-                    // syncing.
+                    // We shouldn't get invalid data from the storage.
+                    // If we do, it's a bug.
+                    assert_eq!(*block.hash(), block_hash, "Block hash mismatch.");
                     trace!(%block_hash, "Linear block found in the local storage.");
-                    // If we found the linear block in the storage it means we should have all of
-                    // its parents as well. If that's not the case then we have a bug.
-                    effect_builder
-                        .immediately()
-                        .event(move |_| Event::StartDownloadingDeploys)
+                    // If we found block in our local storage when syncing trusted hash
+                    // it means we have all of its parents as well (if not then that's a bug that
+                    // will pop up elsewhere). We can start downloading deploys
+                    // starting from the child of _this_ block.
```

Comment:
If I understand correctly the reason why we can do this is not only that we have _blocks_ `0` to `block.height()` on disk, but, more importantly, that we have those blocks' _global states_?

Reply:
That's the assumption, yes.

Reply:
What if there was a previous attempt to sync, that was aborted while downloading the trusted hash's ancestors, but before executing them. Won't we have the _blocks_ in storage, but not the required global states?

In particular, if you start, abort, and start again quickly with the same trusted hash, the first attempt will download and store the trusted block (but nothing else, and without executing), and the second attempt will conclude that we must have everything up to the trusted block?

Reply:
Good point. One solution to this would be to hold onto the block and insert it with its deploys in one transaction but this is not possible with the current storage.